docs/TypedMongo.md

```diff
@@ -0,0 +1,446 @@
+
+## Core types
+
+One of the improvements introduced by `commons-mongo` over the raw Reactive Streams driver are well typed
+representations for field references, query documents, projections, update documents, sort orders, indices etc.
+The raw API uses untyped strings or BSON objects in these situations.
+
+Let's assume some example entities to be used throughout this section:
+
+```scala
+case class MyEntity(
+  id: String,
+  int: Int,
+  str: String,
+  strOpt: Option[String],
+  intList: List[Int],
+  strMap: Map[String, String],
+  data: MyData
+) extends MongoEntity[String]
+object MyEntity extends MongoEntityCompanion[MyEntity]
+
+case class MyData(
+  number: Double,
+  flag: Boolean
+)
+object MyData extends MongoDataCompanion[MyData]
+
+@flatten sealed trait MyUnionEntity extends MongoEntity[String]
+sealed trait HasInt extends MyUnionEntity {
+  def value: Int
+}
+case class FirstCase(id: String) extends MyUnionEntity
+case class SecondCase(id: String, value: Int) extends HasInt
+case class ThirdCase(id: String, value: Int, data: MyData) extends HasInt
+object MyUnionEntity extends MongoEntityCompanion[MyUnionEntity]
+```
+
+### `MongoPropertyRef`
+
+A `MongoPropertyRef[E, T]` is a reference to a property of type `T` inside a document of type `E`.
+It may be a direct reference to a field or a more complex path referring to deeply nested parts of the document.
+A `MongoPropertyRef` is usually obtained with the `.ref` macro which interprets a Scala-level lambda expression into
+a property reference:
+
+```scala
+val IntRef: MongoPropertyRef[MyEntity, Int] = MyEntity.ref(_.int)
+```
+
+Using the `.ref` macro, you can:
+
+* refer to case class fields, e.g. `MyEntity.ref(_.int)`
+* refer to values inside option-like fields, e.g. `MyEntity.ref(_.strOpt.get)`
+* refer to array elements, e.g. `MyEntity.ref(_.intList(2))`
+* refer to map entries, e.g. `MyEntity.ref(_.strMap("key"))`
+* refer to common fields in a sealed hierarchy, e.g. `MyUnionEntity.ref(_.id)`
+* refer to fields of specific case classes in a sealed hierarchy, e.g. `MyUnionEntity.ref(_.as[ThirdCase].data)`
+* refer to common fields of intermediate subtraits in a sealed hierarchy, e.g. `MyUnionEntity.ref(_.as[HasInt].value)`
+
+You can freely combine all the above constructs into more deeply nested references.
+For more examples, see the Scaladoc of the `.ref` macro or 
+[tests](https://github.com/AVSystem/scala-commons/blob/mongo-api/commons-mongo/src/test/scala/com/avsystem/commons/mongo/typed/MongoRefTest.scala).
+
+**NOTE**: some of the above references (e.g. unwrapping optional fields or narrowing sealed hierarchies) introduce
+an _implied filter_ that is automatically included into the query if this reference is used in a MongoDB query document
+or projection.
+
+`MongoPropertyRef` is a fundamental type that serves as an entry point for creating queries, projections, updates,
+sort orders, indices, etc.
+
```

Comment:
Would be nice to mention `MongoRef` and `MongoDataRef`, even if just to signal differences with this type.

Reply:
These differences are explained in scaladoc. I don't want the tutorial to describe too many rarely used details like `MongoRef` and `MongoDataRef`. These don't really appear in everyday-use API.